client/js/lounge.js

```diff
@@ -52,6 +52,84 @@ $(function() {
 
 	var favicon = $("#favicon");
 
+	var emoji = new window.EmojiConvertor();
+	emoji.replace_mode = "unified";
+	emoji.replace_colons("");  // this is needed to create the colons map
+	var emojies = Object.keys(emoji.map.colons);
+
+	// Autocompletion Strategies
+
+	var emojiStrategy = {
+		id: "emoji",
+		match: /\B:([-+\w]*)$/,
+		search: function(term, callback) {
+			callback($.map(emojies, function(e) {
+				return e.indexOf(term) === 0 ? e : null;
+			}));
+		},
+		template: function(value) {
+			return emoji.replace_colons(":" + value + ":") + " " + value;
+		},
+		replace: function(value) {
+			return emoji.replace_colons(":" + value + ":") + " ";
+		},
+		index: 1
+	};
+
+	var nicksStrategy = {
+		id: "nicks",
+		match: /\B@([a-zA-Z_[\]\\^{}|`@][a-zA-Z0-9_[\]\\^{}|`-]*)$/,
+		search: function(term, callback) {
+			if (term[0] === "@") {
+				callback(completeNicks(term.slice(1)).map(function(val) {
+					return "@" + val;
+				}));
+			} else {
+				callback(completeNicks(term));
+			}
+		},
+		template: function(value) {
+			if (value[0] === "@") {
+				return value;
+			}
+			return "@" + value;
+		},
+		replace: function(value) {
+			return value;
+		},
+		index: 1
+	};
+
+	var chanStrategy = {
+		id: "chans",
+		match: /\B(#|\+|&|![A-Z0-9]{5})([^\x00\x0A\x0D\x20\x2C\x3A]+(:[^\x00\x0A\x0D\x20\x2C\x3A]*)?)$/,
```

Comment:
This appears to only filter based on the first letter. If I have 2 channels called "#pokemon" and "#pkmn", and I type "#po" then "#pkmn" will still be at the top of the list.